### PR TITLE
Add support for deploying flows, interpolating variables on import, and refactor into packages a bit

### PR DESCRIPTION
**Pull request by orospakr**
Contributed by https://github.com/joistapp!  Let me know of any copyright/CLA requirements!
- Now forced to use production Heroku auth app even in `dev` mode, because staging environment no longer exists;
- When using `force import` on a project including flows (Flow Builder and Process Builder),the current environment state is used to  determine if any existing Active flows would be overwritten (thus violating the Salesforce requirement that flows cannot be overwritten once being marked active);
- Create support for an `environment.json` config file in the root of a source directory being used for an import, that allows for  interpolating string values in.  This is useful for working around places where environment-specific references (eg., user Ids where Roles/Groups/Profiles cannot be used);
- Created a `project` subpackage, and moved the code responsible for determining the source directory and enumeratings its contents  there;
- Moved all of the API logic into a `salesforce` subpackage, leaving only the CLI components in the root level `main` package.  This enables the other subpackages to import the logic (`main` packages
  are not importable in Go);
- Refactored api version logic; global state is no longer used.

I very much apologise for this god-PR.  In hindsight there's a few tactical decisions I would have made differently.  If this is a major issue, I can take more time to break it up!

Shameless plug: [Joist is hiring!](http://www.joist.com/careers)

### CODE REVIEW

**Comment by orospakr** [comment]
Fixing CI build error.  I see what happened.

**Comment by dcarroll** [comment]
A dependent package needs go 1.6 so I just today modified the yaml file to request that version of go.

**Comment by orospakr** [comment]
@dcarroll sounds like that's a separate error.  seems like the new `util` package I created needs at least one test to pass CI?

**Comment by ViViDboarder** [comment]
Sorry, I know this is tedious, but it's really hard to make sense of 59 file changes and ~2k line changes in one diff. I'm going to try to dig through anyway, but if there are even a couple splits you can make, it would help get this through faster.

Since this is a very large refactor, and I am also planning a refactor of the push.go logic, I think it's worth doing a proper review.

The repackaging of files is a pretty large change itself with a possibility of causing breakage, but I don't think it will be easy to split that out since you probably made other changes already that deviate. But yea, if there's anything easy, splitting would help!

File: aura.go [hunk 1]
```diff
@@ -94,14 +97,14 @@ func runAura(cmd *Command, args []string) {
...
-		ErrorAndExit("force aura create not yet implemented")
+		util.ErrorAndExit("force aura create not yet implemented")
 
 	case "delete":
 		runDeleteAura()
 	case "list":
 		bundles, err := force.GetAuraBundlesList()
 		if err != nil {
-			ErrorAndExit(err.Error())
+			util.ErrorAndExit(err.Error())
```
Comment (ViViDboarder) [inline aura.go]: Kinda general, but maybe putting errors in their own package would be helpful. Maybe not.

File: active.go [hunk 2]
```diff
@@ -35,9 +38,9 @@ func init() {
 
 func runActive(cmd *Command, args []string) {
 	if account == "" {
-		account, _ := Config.Load("current", "account")
-		data, _ := Config.Load("accounts", account)
-		var creds ForceCredentials
+		account, _ := salesforce.Config.Load("current", "account")
```
Comment (ViViDboarder) [inline active.go]: Why is `Config` in `salesforce` rather than `util`? It doesn't actually communicate with the API, but rather loads values from a `Config` file that will then be used to initialize clients.

File: salesforce/packagebuilder.go [hunk 3]
```diff
@@ -289,3 +296,16 @@ func getMetaForPath(path string) (metaName string, objectName string) {
 	objectName = fileName
 	return
 }
+
+func (files *ForceMetadataFilesForType) MetaType() MetaType {
```
Comment (ViViDboarder) [inline salesforce/packagebuilder.go]: What is this? Could use a docstring, not very apparent. Also, it looks like it's a function for a struct, but not sure where that struct lives. Why include it in this file? It doesn't look like it has anything to do with packagebuilder.go.

**Comment by orospakr** [comment]
Will address code review soon, been heads down on a release that went sideways on our project. 😿

File: aura.go (hunk 1 above)
Comment (orospakr) [inline aura.go]: I think that's a fair idea, but let's leave that for another PR.  Just getting it until the util package is a good start, I think! :)

File: active.go (hunk 2 above)
Comment (orospakr) [inline active.go]: Good catch!

Comment (orospakr) [inline active.go]: Addressed.

**Comment by orospakr** [comment]
Self review: Right now the project-level config feature does not account for orgs that have multiple sandboxes on same pod.

**Comment by orospakr** [comment]
Addressed my self review: now you can match your environment to your login with a more flexible regime of regexes on either username or pod hostname.  That should accommodate basically all cases.

**Comment by dcarroll** [comment]
I'd like to merge this into a different branch before promoting to main, that ok with you?

**Comment by orospakr** [comment]
Sure! Sounds great to me. Are you creating an unstable/develop/integration-style branch? :)

**Comment by cwarden** [comment]
I'm going to close this out, @orospakr.  There's some good stuff in here, but it needs to be split into smaller changes.  It also conflicts with the work to split up the code into multiple packages started in 33d6613a71be22987aa60be588e362b85da7cfbc.
